src/substrait/textplan/converter/PlanPrinterVisitor.cpp

```diff
@@ -7,58 +7,567 @@
+      error_listener_->addError(
+          "Literal of this type not yet supported: " +
+          literal.ShortDebugString());
+      return std::string("UNSUPPORTED_LITERAL_TYPE");
+    case ::substrait::proto::Expression::Literal::LITERAL_TYPE_NOT_SET:
+      error_listener_->addError(
+          "Type not specified in literal: " + literal.ShortDebugString());
+      return std::string("UNSPECIFIED_LITERAL_TYPE");
+  }
+  if (literal.nullable()) {
+    text << "?";
+  }
+  return text.str();
+}
+
+std::any PlanPrinterVisitor::visitFieldReference(
+    const ::substrait::proto::Expression::FieldReference& ref) {
+  // TODO -- Move this logic into visitDirectReference and visitMaskedReference.
+  switch (ref.reference_type_case()) {
+    case ::substrait::proto::Expression::FieldReference::kDirectReference:
+      return lookupFieldReference(
+          ref.direct_reference().struct_field().field());
+    case ::substrait::proto::Expression::FieldReference::kMaskedReference:
+      error_listener_->addError(
+          "Masked reference not yet supported: " + ref.ShortDebugString());
+      return "MASKED_REF_NOT_SUPPORTED";
+    case ::substrait::proto::Expression::FieldReference::REFERENCE_TYPE_NOT_SET:
+      error_listener_->addError("Reference type missing from field reference.");
+      return "";
+  }
+  error_listener_->addError(
+      "New field reference type not yet supported: " + ref.ShortDebugString());
+  return "FIELD_REF_TYPE_NOT_SUPPORTED";
+}
+
+std::any PlanPrinterVisitor::visitScalarFunction(
+    const ::substrait::proto::Expression::ScalarFunction& function) {
+  std::stringstream text;
+  text << lookupFunctionReference(function.function_reference()) << "(";
+  // TODO -- Eventually handle options.
+  bool hasPreviousText = false;
+  for (const auto& arg : function.arguments()) {
+    if (hasPreviousText)
+      text << ", ";
+    switch (arg.arg_type_case()) {
+      case ::substrait::proto::FunctionArgument::kEnum:
+        error_listener_->addError(
+            "Enum arguments not yet supported in scalar functions: " +
+            arg.ShortDebugString());
+        text << "ENUM_NOT_SUPPORTED";
+        break;
+      case ::substrait::proto::FunctionArgument::kType:
+        text << ANY_CAST(std::string, visitType(arg.type()));
+        break;
+      case ::substrait::proto::FunctionArgument::kValue:
+        text << ANY_CAST(std::string, visitExpression(arg.value()));
+        break;
+      case ::substrait::proto::FunctionArgument::ARG_TYPE_NOT_SET:
+        error_listener_->addError("Argument type missing the type field.");
+        text << "SCALAR_FUNCTION_ARGUMENT_TYPE_MISSING";
+        break;
+    }
+    hasPreviousText = true;
+  }
+  for (const auto& arg : function.args()) {
+    if (hasPreviousText)
+      text << ", ";
+    text << ANY_CAST(std::string, visitExpression(arg));
+    hasPreviousText = true;
+  }
+  // TODO -- Determine if the output type can be automatically determined.
```

Comment:
Why wouldn't we be able to print the return type?

Reply:
The return type can (and was printed) at one point.  But it was verbose.  I'm wondering if it is possible to divine the type automatically which means we wouldn't need to expose it.

Reply:
Got it.  From the TODO comment I thought there was some trick that made `output_type()` incorrect in some situations.  However, your clarification suggests this is just disabled for brevity.